04 - Functional Programming/04 - Functional Programming.ipynb

```diff
@@ -0,0 +1,3551 @@
+        "\n",
+        "За да демонстрираме разликата, ще създадем функция, която повдига число на квадрат, и ще я приложим върху списък от 40 000 000 числа."
+      ],
+      "metadata": {
+        "id": "TP_Xktyy0GaH"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "from time import time\n",
+        "\n",
+        "def power_up(a):\n",
+        "    return a ** 2\n",
+        "\n",
+        "numbers = list(range(1, 40000000))\n",
+        "\n",
+        "for_loop_start = time()\n",
+        "\n",
+        "for_loop_result = []\n",
+        "for number in numbers:\n",
+        "    for_loop_result.append(power_up(number))\n",
+        "\n",
+        "for_loop_end = time()\n",
+        "print(f'For-looping took {for_loop_end - for_loop_start:.2f} seconds')\n",
+        "\n",
+        "\n",
+        "map_start = time()\n",
+        "\n",
+        "map_result = list(map(power_up, numbers))\n",
+        "\n",
+        "map_end = time()\n",
+        "print(f'map() took {map_end - map_start:.2f} seconds')\n",
+        "\n",
+        "\n",
+        "list_comprehension_start = time()\n",
+        "\n",
+        "list_comprehension_result = [power_up(x) for x in numbers]\n",
+        "\n",
+        "list_comprehension_end = time()\n",
+        "print(f'List comprehension took {list_comprehension_end - list_comprehension_start:.2f} seconds')"
+      ],
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "mfnNOi640c7h",
+        "outputId": "e7e2987d-0c6c-4702-bf02-d6e6c9b88405"
+      },
+      "execution_count": 74,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "name": "stdout",
+          "text": [
+            "For-looping took 19.72 seconds\n",
+            "map() took 13.60 seconds\n",
+            "List comprehension took 15.08 seconds\n"
+          ]
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Основната разлика във времето идва от факта, че `append()` операцията е бавна. `map` и list comprehension-ите използва хитрини при заделянето на паметта, затова може да се каже че са с около 30% по-бързи. Няма да се впускам повече в темата, за по-любопитните, [тук](https://web.archive.org/web/20190319205826/http://blog.cdleary.com/2010/04/efficiency-of-list-comprehensions/) нещата са обяснение в повече детайли."
+      ],
+      "metadata": {
+        "id": "E5_wupc73MQm"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "На въпросът, кой от трите подхода се използват, моят отговор е следния: list comprehension и `map` дават доста по-добро представяне спрямо `for` loop + `append`, затова и са за предпочитане. Синтаксисът на list comprehension-ите е по-прост спрямо този на `map` (особенно ако трябва и да филтрираме елементите), което води до това, че **list comprehension-ите са за предпочитане**."
```

Comment:
"list comprehension-ите са за предпочитане" not if you have 5 expressions nested inside one another. Readability > Performance.

Reply:
Eh, I could argue about this one - if you have multiple expressions that work with the same data, then it could become a function. Then you can call this function inside the comprehension. 

Otherwise, agree that list comprehensions are not usable if you do multiple things inside the for loop. 

Agree with "Readability > Performance", but will not change the wording.